pkg/memory/short_term.go

```diff
@@ -0,0 +1,159 @@
+// that have already been performed.
+
+package memory
+
+import (
+	"reflect"
+	"sync"
+	"time"
+)
+
+// ShortTermMemoryBuilder builds of short term memory objects.
+//
+type ShortTermMemoryBuilder struct {
+	// How long to remember actions.
+	duration time.Duration
+}
+
+// ShortTermMemory stores a set of items for a given period of time.
+//
+type ShortTermMemory struct {
+	// How long to remember actions.
+	duration time.Duration
+
+	// There will be a cell for each action stored, containing the action itself and the time it was
+	// added to the memory.
+	cells []*ShortTermCell
+
+	// Mutex used to prevent simultaneous updates of the data structures.
+	mutex *sync.Mutex
+}
+
+// ShortTermCell stores each individual action, and the time it was added to the memory.
+//
+type ShortTermCell struct {
+	// The item stored in the cell.
+	item interface{}
+
+	// The time when the cell was created or updated.
+	stamp time.Time
+}
+
+// NewShortTermMemoryBuilder creates a builder that can create short term memory objects.
+//
+func NewShortTermMemoryBuilder() *ShortTermMemoryBuilder {
+	b := new(ShortTermMemoryBuilder)
+	return b
+}
+
+// Duration sets how long objects in the memory will be remembered. The default is zero, which means
+// that objects won't be remembered at all.
+//
+func (b *ShortTermMemoryBuilder) Duration(duration time.Duration) *ShortTermMemoryBuilder {
+	b.duration = duration
+	return b
+}
+
+// Build creates a new short term memory object with the configuration stored in the builder.
+//
+func (b *ShortTermMemoryBuilder) Build() (m *ShortTermMemory, err error) {
+	m = new(ShortTermMemory)
+	m.duration = b.duration
+	m.cells = make([]*ShortTermCell, 0)
+	m.mutex = &sync.Mutex{}
+	return
+}
+
+// Add adds a new item to the memory.
+//
+func (m *ShortTermMemory) Add(item interface{}) {
+	m.mutex.Lock()
+	defer m.mutex.Unlock()
+
+	cell := m.findMatchingCell(item)
+	if cell == nil {
+		cell = new(ShortTermCell)
+		cell.item = item
+		m.cells = append(m.cells, cell)
+	}
+	cell.stamp = time.Now()
+}
+
+func (m *ShortTermMemory) Has(item interface{}) bool {
+	// Purge cells before checking.
+	m.CleanShortTermMemory()
+
+	m.mutex.Lock()
+	defer m.mutex.Unlock()
+
+	return m.findMatchingCell(item) != nil
+}
+
+// Len returns the number of items inside the memory.
+//
+func (m *ShortTermMemory) Len() int {
+	m.CleanShortTermMemory()
+
+	m.mutex.Lock()
+	defer m.mutex.Unlock()
+
+	return len(m.cells)
+}
+
+// PurgeExpiredCells finds the aged cells and removes them.
+//
+func (m *ShortTermMemory) purgeExpiredCells() {
+	now := time.Now()
+	for idx, cell := range m.cells {
+		age := now.Sub(cell.stamp)
+		// purge cells in place.
+		if age >= m.duration {
+			// zeroing the value of the cell so that it wouldn't be refrenced by the underlying array
+			// causing GO's grabage collector to collect the allocated memory.
+			cell = nil
+			m.cells = append(m.cells[:idx], m.cells[idx+1:]...)
```

Comment:
I think this logic is correct. Maybe a bit hard to understand. Would be nice to have a comment explaining that cells are ordered from oldest to newest.

Also take into account that when you do `cell = nil` what you are clearing is the `cell` loop variable, but you aren't touching the underlying array. If you really want to clear the item in the array you will need to do `m.cells[idx] = nil`.